daemon/api_snaps.go

```diff
@@ -1016,3 +1032,55 @@ func snapUnholdMany(inst *snapInstruction, st *state.State) (res *snapInstructio
 		Tasksets: tss,
 	}, nil
 }
+
+func snapSetupFromChange(chg *state.Change) (*snapstate.SnapSetup, error) {
+	for _, t := range chg.Tasks() {
+		// Check a known task of each change that we know keep snap info.
+		if t.Kind() != "prerequisites" {
+			continue
+		}
+		return snapstate.TaskSnapSetup(t)
+	}
+	return nil, nil
+}
+
+// verifySnapdCanRefresh verifies that the system is in a state where snapd can
+// revert or refresh. Currently this means that no other snap changes
+// are allowed to execute.
+func verifySnapdCanRefresh(st *state.State) error {
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		// XXX: Currently limited to any snap changes (i.e removal, refresh revert) however
+		// this should maybe be extended to all changes?
+		snapsup, err := snapSetupFromChange(chg)
+		if err != nil {
+			return err
+		} else if snapsup != nil {
+			return fmt.Errorf("cannot revert or refresh snapd while other changes are executing")
+		}
+	}
+	return nil
+}
+
+// verifySnapdIsNotRefreshing verifies that no change is currently running
+// for a snapd refresh/revert. While snapd is updating we do not allow other
+// snap changes to be executing, to avoid any incompatibilities in the task system,
+// between snapd versions.
+func verifySnapdIsNotRefreshing(st *state.State) error {
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		// XXX: Currently limited to any snap changes (i.e removal, refresh revert) however
+		// this should maybe be extended to all changes?
+		snapsup, err := snapSetupFromChange(chg)
+		if err != nil {
+			return err
+		} else if snapsup != nil && snapsup.SnapName() == "snapd" {
+			return fmt.Errorf("cannot do changes while snapd is refreshing/reverting")
```

Comment:
need some input on this error message

Reply:
this seems too much of blocking, we really want to block only going to older rversions, anyway as I said the check need to be done differently/elsewhere

Reply:
now I remember that we discussed this and the idea was that it doesn't make sense to make a request if the snapd version is not determined

Reply:
I had it noted in my notes that we ended up agreeing that it we would block while snapd was updating (but I most likely just failed to note it was during any *downgrade* not *upgrade*). 

I updated the logic to just block during downgrades, which makes sense. Logic changes to task system should always be backwards-compatible, so the direction where we cannot plan is the downgrade case.